src/features/island/buildings/components/building/composters/AdvancedComposter.tsx

```diff
@@ -0,0 +1,132 @@
+import React, { useContext, useState } from "react";
+
+import advancedComposter from "assets/composters/composter_advanced.png";
+import advancedComposterClosed from "assets/composters/composter_advanced_closed.png";
+import { PIXEL_SCALE } from "features/game/lib/constants";
+import { Context } from "features/game/GameProvider";
+import { useActor, useInterpret, useSelector } from "@xstate/react";
+import {
+  CompostingContext,
+  MachineInterpreter,
+  composterMachine,
+  MachineState,
+} from "features/island/buildings/lib/composterMachine";
+import { ComposterModal } from "./ComposterModal";
+import { SUNNYSIDE } from "assets/sunnyside";
+
+const isIdle = (state: MachineState) => state.matches("idle");
+const isComposting = (state: MachineState) => state.matches("composting");
+const isReady = (state: MachineState) => state.matches("ready");
+
+export const AdvancedComposter: React.FC = () => {
+  const { gameService } = useContext(Context);
+  const [gameState] = useActor(gameService);
+  const [showModal, setShowModal] = useState(false);
+
+  const composter =
+    gameState.context.state.buildings["Advanced Composter"]?.[0];
+
+  const composterMachineContext: CompostingContext = {
+    gameService,
+    readyAt: composter && composter.producing?.readyAt,
+  };
+
+  const composterService = useInterpret(composterMachine, {
+    context: composterMachineContext,
+  }) as unknown as MachineInterpreter;
+
+  const idle = useSelector(composterService, isIdle);
+  const composting = useSelector(composterService, isComposting);
+  const ready = useSelector(composterService, isReady);
+
+  const [
+    {
+      context: { secondsTillReady },
+    },
+  ] = useActor(composterService);
+
+  const startComposter = () => {
+    composterService.send({
+      type: "START_COMPOST",
+      event: "composter.started",
+      buildingId: composter!.id,
+      building: "Advanced Composter",
+    });
+  };
+
+  const handleCollect = () => {
+    composterService?.send({
+      type: "COLLECT",
+      event: "compost.collected",
+      buildingId: composter!.id,
+      building: "Advanced Composter",
+    });
+  };
+
+  const handleClick = () => {
+    if (idle || composting) {
+      // composterAudio.play();
+      setShowModal(true);
+      return;
+    }
+
+    if (ready) {
+      handleCollect();
+      return;
+    }
+  };
+
+  return (
+    <>
+      <div
+        className="absolute"
+        style={{
+          width: `${PIXEL_SCALE * 27}px`,
+          bottom: `${PIXEL_SCALE * 0}px`,
+          left: `${PIXEL_SCALE * 2}px`,
+        }}
+        onClick={handleClick}
+      >
+        <img
+          src={idle ? advancedComposter : advancedComposterClosed}
+          style={{
+            width: `${PIXEL_SCALE * 27}px`,
+            bottom: `${PIXEL_SCALE * 0}px`,
+          }}
+          className="absolute"
+          alt="Advanced Composter"
+        />
+      </div>
+      <ComposterModal
+        {...{
```

Comment:
I think it would be better to actually get the state inside of this Modal rather than passing it down as props. You're also passing the whole state but then passing the inventory separately.

I think just read from the machine directly inside here to limit the re-renders.